cmake: Add fail pattern to `try_append_cxx_flags`

hebasto opened the pull request:
This PR addresses scenarios where the build environment, such as OSS-Fuzz, provides the `-Wno-error=unknown-warning-option` flag, which undermines our use of `-Werror` when checking compiler flags.

To reproduce the issue on the master branch @ c1f0a89d9caeb62a0a5f3462b454746941e626e6:
```
$ env CXX=clang++ CC=clang CXXFLAGS="-Wno-error=unknown-warning-option" cmake -B build 2>&1 | grep WDUPLICATED_BRANCHES
-- Performing Test CXX_SUPPORTS__WDUPLICATED_BRANCHES
-- Performing Test CXX_SUPPORTS__WDUPLICATED_BRANCHES - Success
```

This output shows that the `-Wduplicated-branches` flag is incorrectly treated as recognised by the compiler.

Noticed while working on the [migration](https://github.com/google/oss-fuzz/pull/14457) to Ubuntu 24.04 in OSS-Fuzz.

maflcko commented on cmake/module/TryAppendCXXFlags.cmake:
```diff
@@ -64,7 +64,7 @@ function(try_append_cxx_flags flags)
   set(CMAKE_REQUIRED_FLAGS "${flags_as_string} ${working_compiler_werror_flag}")
   set(compiler_result CXX_SUPPORTS_${id_string})
-  check_cxx_source_compiles("${source}" ${compiler_result})
+  check_cxx_source_compiles("${source}" ${compiler_result} FAIL_REGEX [=[\\[-Wunknown-warning-option\\]]=])
```
> This output shows that the `-Wduplicated-branches` flag is incorrectly treated as recognised by the compiler.

Yes, but this is what the user asked for by setting `-Wno-error=unknown-warning-option`?

I don't think we should add code to silently undo switches that the user turned on.

Maybe I am missing something obvious, but I don't see the risk or the downside of just leaving this as-is. 


> Noticed while working on the https://github.com/google/oss-fuzz/pull/14457 to Ubuntu 24.04 in OSS-Fuzz.

It seems unrelated to the migration, so I think it would be best to drop it there.

hebasto committed:
cmake: Add fail pattern to `try_append_cxx_flags`

This change addresses scenarios where the build environment, such as
OSS-Fuzz, provides the `-Wno-error=unknown-warning-option` flag, which
undermines our use of `-Werror` when checking compiler flags.

hebasto commented on cmake/module/TryAppendCXXFlags.cmake:
```diff
@@ -64,7 +64,7 @@ function(try_append_cxx_flags flags)
   set(CMAKE_REQUIRED_FLAGS "${flags_as_string} ${working_compiler_werror_flag}")
   set(compiler_result CXX_SUPPORTS_${id_string})
-  check_cxx_source_compiles("${source}" ${compiler_result})
+  check_cxx_source_compiles("${source}" ${compiler_result} FAIL_REGEX [=[\\[-Wunknown-warning-option\\]]=])
```
Thanks! An inline explanatory comment has been added.

maflcko commented:
Maybe it can be put into draft for now or closed?
